File: leetcode/716. Max Stack.py
from sortedcontainers import SortedDict

class Node:
    def __init__(self, val=None):
        self.val = val
        self.next = None
        self.last = None
        

class MaxStack:

    def __init__(self):
        self.dic = SortedDict()
        self.root = Node()
        self.root.last, self.root.next = self.root, self.root
        

    def push(self, x: int) -> None:
        if x not in self.dic:
            self.dic[x] = []
        node = Node(x)
        self.root.next.last, self.root.next, node.last, node.next = node, node, self.root, self.root.next
        self.dic[x].append(node)
        

    def pop(self) -> int:
        node = self.root.next
        node.next.last, node.last.next = node.last, node.next
        node_lst = self.dic[node.val]
        node_lst.pop()
        if len(node_lst) == 0:
            self.dic.pop(node.val)
        return node.val
    

    def top(self) -> int:
        return self.root.next.val
        

    def peekMax(self) -> int:
        return self.dic.peekitem()[0]


    def popMax(self) -> int:
        val, node_lst = self.dic.peekitem()
        node = node_lst.pop()
        if len(node_lst) == 0:
            self.dic.pop(val)
        node.next.last, node.last.next = node.last, node.next
        return val


# Your MaxStack object will be instantiated and called as such:
# obj = MaxStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.peekMax()
# param_5 = obj.popMax()

File: leetcode/716. Max Stack.rs
/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
struct MaxStack {
    values: Vec<i32>,
}

impl MaxStack {

    fn new() -> Self {
        MaxStack {
            values: Vec::new(),
        }
    }
    
    fn push(&mut self, x: i32) {
        self.values.push(x);
    }
    
    fn pop(&mut self) -> i32 {
        self.values.pop().unwrap()
    }
    
    fn top(&self) -> i32 {
        *self.values.last().unwrap()
    }
    
    fn peek_max(&self) -> i32 {
        *self.values.iter().max().unwrap()
    }
    
    fn pop_max(&mut self) -> i32 {
        let max = self.peek_max();
        for i in (0..self.values.len()).rev() {
            if self.values[i] == max {
                return self.values.remove(i)
            }
        }
        unimplemented!()
    }
}

/**
 * Your MaxStack object will be instantiated and called as such:
 * let obj = MaxStack::new();
 * obj.push(x);
 * let ret_2: i32 = obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: i32 = obj.peek_max();
 * let ret_5: i32 = obj.pop_max();
 */
